lib/Dialect/Moore/Transforms/LowerConcatRef.cpp

```diff
@@ -0,0 +1,126 @@
+//===- LowerConcatRef.cpp - moore.concat_ref lowering ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the LowerConcatRef pass.
+// It's used to disassemble the moore.concat_ref. Which is tricky to lower
+// directly. For example, disassemble "{a, b} = c" onto "a = c[7:3]"
+// and "b = c[2:0]".
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Moore/MooreOps.h"
+#include "circt/Dialect/Moore/MoorePasses.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_LOWERCONCATREF
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+using namespace mlir;
+
+namespace {
+
+template <typename OpTy>
+struct ConcatRefLowering : public OpConversionPattern<ConcatRefOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ConcatRefOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    // Use to collect the operands of concatRef.
+    SmallVector<Value> values;
+    for (auto operand : op.getValues()) {
+      // Handle the nested concatRef, and then collect its operands.
+      if (auto *defOp = operand.getDefiningOp(); isa<ConcatRefOp>(defOp)) {
+        for (auto nestedOperand : defOp->getOperands())
+          values.push_back(nestedOperand);
+        rewriter.eraseOp(defOp);
```

Comment:
I think it's illegal to delete defOp at this point

Reply:
I'm not sure. But I think you're right!